src/MilestoneTG.Extensions.Configuration.S3/S3ConfigurationProvider.cs

```diff
@@ -62,30 +70,38 @@ public S3ConfigurationProvider(S3ConfigurationSource source)
         /// </summary>
         public void WaitForReloadToComplete(TimeSpan timeout)
         {
-            reloadTaskEvent.WaitOne(timeout);
+            this.reloadTrigger.BlockThreadUntilTriggered(timeout);
         }
 
         /// <summary>
-        ///  Loads (or reloads) the data for this provider.
+        /// Loads (or reloads) the data for this provider.
         /// </summary>
         public override void Load() => Load(false);
-
+        
         private void Load(bool reload) => LoadAsync(reload).ConfigureAwait(false).GetAwaiter().GetResult();
-
+        
         private async Task LoadAsync(bool reload)
         {
             try
             {
-                using (GetObjectResponse s3Response = await s3.GetObjectAsync(source.BucketName, source.Key).ConfigureAwait(false))
+                GetObjectRequest request = new GetObjectRequest()
                 {
-                    Data = await source.Parser.ParseAsync(s3Response).ConfigureAwait(false);
-                }
+                    BucketName = config.BucketName,
+                    Key = config.Key,
+                    EtagToNotMatch = previousEtag
+                };
 
+                using (GetObjectResponse s3Response = await s3client.GetObjectAsync(request).ConfigureAwait(false))
+                {
+                    Data = await parser.ParseAsync(s3Response).ConfigureAwait(false);
+                    this.previousEtag = s3Response.ETag;
+                }
+                
                 OnReload();
```

Comment:
This still isn't quire right for the performance yo're looking to gain. You need to check that the response has content or not (ContentLength == 0) and only change Data and reload if there was content.

Like this...

```cs
                using (GetObjectResponse s3Response = await s3client.GetObjectAsync(request).ConfigureAwait(false))
                {
                    if (s3Response.ContentLength > 0)
                    {
                        Data = await parser.ParseAsync(s3Response).ConfigureAwait(false);
                        this.previousEtag = s3Response.ETag;
                        OnReload();
                    }
                }
```

Reply:
I went ahead and made the necessary performance enhancement on master and pushed version 1.1 to Nuget.

Reply:
According to my experiment and the AWS .NET SDK documentation, the AmazonS3Client.GetObjectAsync(GetObjectRequest, CancellationToken) method throws AmazonS3Exception (and it is handled by the catch clause) if the file did not change since the last request was made (ETag must be provided to do so). Therefore the statements following the invocation of GetObjectAsync() will not be executed, therefore there's no need to check the GetObjectResponse.ContentLength property.

Another possible way to do this is to perform a HEAD request with the ETag, then check the HTTP status code, and if it is not NotModified, then perform the GetObjectAsync request. However, in this case, two HTTP request is made, which is more expensive than handling the exception. That's why I choose my original approach.

Reply:
> I went ahead and made the necessary performance enhancement on master and pushed version 1.1 to Nuget.

Thank you for your effort. Unfortunately, the file is still being downloaded by each poll. According to the current revision on your master branch
`GetObjectRequest request = new GetObjectRequest()
{ 
	BucketName = source.BucketName,
	Key = source.Key,
	EtagToNotMatch = previousEtag
};

using (GetObjectResponse s3Response = await s3.GetObjectAsync(source.BucketName, source.Key).ConfigureAwait(false))`
you forgot to adjust the arguments of GetObjectAsync to use the request object created above. And if you adjust the code, you don't need the IF statement to check ContentLength (due to the reason I mentioned above).

Reply:
I made the fix and changed the code to check the metadata first on reload as you suggested. It's disappointing Amazon didn't fix that (even though this thread says otherwise (https://forums.aws.amazon.com/thread.jspa?threadID=77995&tstart=0)). 

Nuget package 1.2.0 has been published.

Reply:
Thank you very much! I reviewed your changes and I agree with them. I'm glad that the issue is solved, I really appreciate it. Your code will save us a lot of time.

I think this PR can be deleted or marked as "declined".